pkg/controller/topology_test.go

```diff
@@ -908,42 +910,112 @@ func TestTopologyAggregation(t *testing.T) {
 				{Segments: map[string]string{"com.example.csi/zone": "zone1"}},
 			},
 		},
-		"random node: missing node info": {
-			nodeLabels:        []map[string]string{{}, {}, {}},
-			topologyKeys:      nil,
+		"random node: no nodes": {
+			nodeLabels: nil,
+			topologyKeys: []map[string][]string{
+				{testDriverName: []string{"com.example.csi/zone"}},
+				{testDriverName: []string{"com.example.csi/zone"}},
+				{testDriverName: []string{"com.example.csi/zone"}},
+			},
 			expectedRequisite: nil,
+			expectError:       true,
 		},
-		"selected node: missing node info": {
-			hasSelectedNode:   true,
-			nodeLabels:        []map[string]string{{}, {}, {}},
+		"random node: missing matching node info": {
+			nodeLabels: []map[string]string{
+				{"com.example.csi/foo": "bar"},
+			},
+			topologyKeys: []map[string][]string{
+				{testDriverName: []string{"com.example.csi/zone"}},
+				{testDriverName: []string{"com.example.csi/zone"}},
+				{testDriverName: []string{"com.example.csi/zone"}},
+			},
+			expectedRequisite: nil,
+			expectError:       true,
+		},
+		// Node has not been upgraded yet
+		"random node: no CSINodes": {
+			nodeLabels: []map[string]string{
+				{"com.example.csi/zone": "zone1"},
+				{"com.example.csi/zone": "zone2"},
+				{"com.example.csi/zone": "zone2"},
+			},
 			topologyKeys:      nil,
 			expectedRequisite: nil,
 		},
+		// Driver on node has not been updated to report topology keys
 		"random node: missing keys": {
```

Comment:
Should we have a test case where there's only a single CSINode with keys, in which there should be no errors?

Reply:
Do the existing "same keys and values across cluster" and "different values across cluster" cases address this?

Reply:
Not quite. I'm thinking about the case where k8s nodes are partially upgraded to 1.14, so some nodes have keys but others don't. Doesn't matter if Node objects have the right topology values.

Reply:
Something like:
```golang
nodeLabels: []map[string]string{
				{"com.example.csi/zone": "zone1"},
				{"com.example.csi/zone": "zone2"},
				{"com.example.csi/zone": "zone2"},
			},
			topologyKeys: []map[string][]string{
				{testDriverName: []string{"com.example.csi/zone"}},
			},
```
in which the requisite should only have "com.example.csi/zone": "zone1", right?

Reply:
Added 2 more test cases.

So actually, we in your case above, requisite has both zone1 and zone2. even though the other nodes don't have csiNode.  It's because csiNode is only used to determine which topology keys to use, but we just iterate over all nodes that have the topology key.